contracts/RFQ.sol

```diff
@@ -0,0 +1,154 @@
+        feeCollector = _newFeeCollector;
+
+        emit SetFeeCollector(_newFeeCollector);
+    }
+
+    function fillRFQ(
+        Offer calldata offer,
+        bytes calldata makerSignature,
+        bytes calldata makerTokenPermit,
+        bytes calldata takerTokenPermit,
+        address payable recipient
+    ) external payable override {
+        _fillRFQ(RFQOrder({ offer: offer, recipient: recipient, feeFactor: 0 }), makerSignature, makerTokenPermit, takerTokenPermit, bytes(""));
+    }
+
+    function fillRFQ(
+        RFQOrder calldata order,
+        bytes calldata makerSignature,
+        bytes calldata makerTokenPermit,
+        bytes calldata takerTokenPermit,
+        bytes calldata takerSignature
+    ) external override {
+        _fillRFQ(order, makerSignature, makerTokenPermit, takerTokenPermit, takerSignature);
+    }
+
+    function _fillRFQ(
+        RFQOrder memory _rfqOrder,
+        bytes memory _makerSignature,
+        bytes memory _makerTokenPermit,
+        bytes memory _takerTokenPermit,
+        bytes memory _takerSignature
+    ) private {
+        Offer memory _offer = _rfqOrder.offer;
+        // check the offer deadline and fee factor
+        if (_offer.expiry < block.timestamp) revert ExpiredOffer();
+        if (_rfqOrder.feeFactor > Constant.BPS_MAX) revert InvalidFeeFactor();
+
+        // check if the offer is available to be filled
+        (bytes32 offerHash, bytes32 rfqOrderHash) = _getRFQOrderHash(_rfqOrder);
+        if (filledOffer[offerHash]) revert FilledOffer();
+        filledOffer[offerHash] = true;
+
+        // check maker signature
+        if (!SignatureValidator.isValidSignature(_offer.maker, getEIP712Hash(offerHash), _makerSignature)) revert InvalidSignature();
+
+        // check taker signature if needed
+        if (_offer.taker != msg.sender) {
+            if (!SignatureValidator.isValidSignature(_offer.taker, getEIP712Hash(rfqOrderHash), _takerSignature)) revert InvalidSignature();
+        }
+
+        // transfer takerToken to maker
+        if (_offer.takerToken.isETH()) {
+            if (msg.value != _offer.takerTokenAmount) revert InvalidMsgValue();
+            Address.sendValue(_offer.maker, _offer.takerTokenAmount);
+        } else {
+            _collect(_offer.takerToken, _offer.taker, _offer.maker, _offer.takerTokenAmount, _takerTokenPermit);
+        }
+
+        // collect makerToken from maker to this
+        _collect(_offer.makerToken, _offer.maker, address(this), _offer.makerTokenAmount, _makerTokenPermit);
+
+        // transfer makerToken to recipient (sub fee)
+        uint256 fee = (_offer.makerTokenAmount * _rfqOrder.feeFactor) / Constant.BPS_MAX;
+        // determine if WETH unwrap is needed, send out ETH if makerToken is WETH
+        address makerToken = _offer.makerToken;
+        if (makerToken == address(weth)) {
+            weth.withdraw(_offer.makerTokenAmount);
+            makerToken = Constant.ETH_ADDRESS;
+        }
```

Comment:
It's mainly for collecting token from maker since ETH cannot be collected, only via send. But usually WETH is a medium and taker wants ETH most of the time.